### ISSUE TITLE
[kohlchan.net]: Support discussion

### ISSUE THREAD

**Issue by sunabozu** [SthephanShinkufag/Dollchan-Extension-Tools#1332]
**Is your feature request related to a problem?**
After the recent switch to [https://gitgud.io/Tjark/KohlNumbra](url) Kohlchan doesn't support Dollchan anymore.

**Describe the solution you'd like to see in the implementation**
Can somebody look into it and tell if it's possible to add the support?

**Comment by StephenLynx** [comment]
I'd just like to point that they are using lynxchan now. But they made a few changes to their front-end, so support for other lynxchan sites might not work with them.

**Comment by SthephanShinkufag** [comment]
The lynxchan engine is developing, now it has a new posting mechanism, which conflicts with the Dollchan. Posting does not work on all lynxchan imageboards, at the moment.

**Comment by StephenLynx** [comment]
Ah, you mean the api merge? Yeah, that changed on 2.2.

**Comment by ganego** [comment]
Things that do not work:  
- Youtube previews (Example: https://nocsp.kohlchan.net/tv/res/3758.html )
- Tab bar: New post warning will not be removed after viewing the tab (green number & favicon).
- "Post menu" button (no idea what this is for).
- Password has to be 8 or less digits. Maybe admin will fix this.

Thank you!

**Comment by sunabozu** [comment]
@ganego The last issue can be easily fixed by setting a shorter password in the settings. But after that you get a new problem:
> This board requires at least one file when creating threads.

It doesn't recognize the attached files.

**Comment by ganego** [comment]
Update: YT preview and links in general can be fixed by setting `this.hasTextLinks = true;`

**Comment by ganego** [comment]
- Admin changed PW length. Now allows for more than 8 chars again.
- Regarding favicons not changing when new posts were viewed: 

There is this code part (es6 version):

```
updateIcon(isError) {
	if(!isError && !newPosts) {
		this._setIcon(this.originalIcon);
	} else if(this._hasIcons) {
		this._setIcon(isError ? this._iconError :
			repliesToYou.size ? this._getIconYou(newPosts) : this._getIconNew(newPosts));
	}
}
```
Now if you add a `console.log(this.originalIcon);` before the first _setIcon it will say `"https://kohlchan.net/favicon.ico"` as expected. But setting the icon does not work. If I replace `this.originalIcon` with `https://google.com/favicon.ico` it works perfectly fine and after viewing the thread, the favicon in the tab bar will switch to the google favicon. Strange.

**Comment by ganego** [comment]
Update: Favicon fixed by admin (was wrong http header).

**Comment by ganego** [comment]
So the only thing left not working is posting/with images.
Posting itself works (without image), but dollchan won't realize this (post is successfully sent and shows up, when refreshing). Cannot really help there.  
One difference between DC and lynxchan autoupdater: DC reloads the whole page, whereas lynxchan just fetches some much smaller json.

**Comment by StephenLynx** [comment]
https://gitgud.io/LynxChan/PenumbraLynx/blob/master/static/js/api.js#L303

**Comment by ganego** [comment]
So I pretty much spammed the source with console.log() to find out what actually happens (functions like ´html5Submit´, ´sendHTML5Post´ and `_initAjaxPosting (the this.form.onsubmit part`, and, uhm, nothing happens?  
I tried it on endchan and pretty much all my debug log() fired, whereas on KC nothing happened.

The first thing that should be triggered would be:

```
_initAjaxPosting() {
	let el;
	console.log("_initAjaxPosting")
	if(aib.qFormRedir && (el = $q(aib.qFormRedir, this.form))) {
		aib.disableRedirection(el);
		console.log("aib.disableRedirection")
	}
	console.log(this.form)
	this.form.onsubmit = e => {
		console.log("this.form.onsubmit - BEFORE")
		$pd(e);
		$popup('upload', Lng.sending[lang], true);
		html5Submit(this.form, this.subm, true).then(checkUpload)
			.catch(err => $popup('upload', getErrorMessage(err)));
		console.log("this.form.onsubmit - AFTER")
	};
}
```

The init works fine (console.log(this.form) gets printed upon page-load and is `<form action="/replyThread.js" enctype="multipart/form-data" method="post" style="display: inline-block; text-align: left;">`).
But the event does not get triggered at all. On endchan it gets triggered (.log() prints something).
Now why does this simply event not get triggered?  

I saw one difference from endchan to kohlchan: `document.getElementById("formButton")` (which is `this.subm`) has two onClick events on KC and only the dollchan one on EC. But I also tested this by removing the whole `#formButton`  and adding it again to remove the KC event listener:

init() of KC:

```
var ff = document.getElementById("formButton");
var hh = ff.outerHTML;
var par = ff.parentElement;
ff.parentElement.removeChild(ff.parentElement.children[2]);
par.innerHTML = par.innerHTML + hh;
```

This removed the eventListener, but it did not trigger the event listener anyway (and I found out, that the original onClick event seems to be the thing still making KC somewhat work).

The fuck?

**Comment by StephenLynx** [comment]
https://gitgud.io/LynxChan/PenumbraLynx/blob/master/static/js/api.js#L86
The method kc uses to enable js on posting is different from the one endchan uses.

**Comment by ganego** [comment]
Ok more progress.
First of all remove the old button and make it "submit".

KC init:
```
init() {
	super.init();
	$script('thread.postReply = function() {}');
	$script('thread.startTimer = function(time) {}');
	
	var ff = document.getElementById("formButton");
	ff.type = "submit";
	var hh = ff.outerHTML;
	var par = ff.parentElement;
	ff.parentElement.removeChild(ff.parentElement.children[2]);
	par.innerHTML = par.innerHTML + hh;
	
	return false;
}
```

Now, fix the api URL:

In `async sendHTML5Post(form, data, needProgress, hasFiles) {` change it to `return $ajax(form.action + '?json=1', ajaxParams)`.

Next error to fix `UnsupportedMediaTypeError: unsupported content-type`. - Working on that.

**Comment by SthephanShinkufag** [comment]
@ganego, I'm already doing before all what you doing now, don't make a work twice.

**Comment by SthephanShinkufag** [comment]
The fix will be soon.

**Comment by StephenLynx** [comment]
@ganego because you have to use multipart to send files. Endchan uses the deprecated api that was just a big binary blob on the request body, while the new api is limited to the no-js input.

**Comment by SthephanShinkufag** [comment]
Seems the posting is working now.
Please tell mes some boards on new api **with captcha**, it need to be fixed.

**Comment by SthephanShinkufag** [comment]
Dollchan sends these parameters when old API:
```
captchaId  : cookieObj.captchaid,
bypassId   : cookieObj.bypass,
auth       : { login: cookieObj.login, hash: cookieObj.hash }
```

But for new API I ignored them. I need a board with captcha to test new API.
I also ignored these parameters for files:
```
          form.append('fileMd5', file.md5);
          form.append('fileMime', file.mime);
          form.append('fileSpoiler', file.spoiler || '');
          form.append('fileName', file.name);
```

Everything works for now, maybe I didn't notice the problems yet.

**Comment by ganego** [comment]
Lynxchan constructor you set `this._hasNewAPI = false;` and in init() you set `this._hasNewAPI = true;`. Any reason for that?

Also: `ajaxParams = { data, method: 'POST' };`. This easy? Now I feel stupid for all the stuff I tried to get it to work (I tried to put dataObj manually in a string with correct boundaries for the multipart/form-data but it would always error with "BadRequestError: stream ended unexpectedly". Oh well, at least I learned new stuff :)

Thank you!

**Comment by SthephanShinkufag** [comment]
Heh. Sometimes decisions are much easier than they seem.
I’m worried about fields that are generated by internal scripts. Bypass, captchaid, file.md5 and so on. Whether they are critically needed or not. And if so, which ones.

**Comment by ganego** [comment]
What does the bypass checkbox mean: Admin answer: "If checked, it checks for bypass first (request permission to post with Tor for example) and only if true sents the post with media to server." & "Checkbox is now removed. It didn't even worked like intended, so nothing missing."  

The rest: No idea, let's wait for @StephenLynx

**Comment by StephenLynx** [comment]
The bypass check just checks if the user will require a block bypass to post before actually trying to post. It saves him time uploading files, mostly. All it does is to use the api on an entirely different operation.

Those parameters fileMd5 and others similars are used to reference files and spoil individual files. If you ignore that, users won't be able to post files without uploading them if they already exist on the server.

**Comment by StephenLynx** [comment]
https://gitgud.io/LynxChan/PenumbraLynx/blob/master/static/js/thread.js#L604
How the check works.

https://gitgud.io/LynxChan/PenumbraLynx/blob/master/static/js/postCommon.js#L325
How the file processing works.

**Comment by ganego** [comment]
Password field in the delete form is not auto-filled.

**Comment by kr4ssi** [comment]
Images don't open in the viewer sometimes, because they have no extension, there's a data-attribute with the correct mime though. Can that be fixed?

**Comment by SthephanShinkufag** [comment]
@kr4ssi 

Need a link to the long-live post with this kind of image.

**Comment by ganego** [comment]
Yeah I observed this bug some time ago too, but it's somewhat rare and I figured you didn't want to add an additional mime check just for some lynxchan fuckup.  
Take this thread for example: https://nocsp.kohlchan.net/prog/res/1564.html

The problem is, the board software sometimes does not add the proper extension, so it will show up as `...imagepng` instead of `...imagepng.png`. DC only checks for the extension after the dot, so it won't be recognized. But lynxchan send a data tag: `document.getElementsByClassName("imgLink")[0].getAttribute("data-filemime")`.

**Comment by SthephanShinkufag** [comment]
This is a kohlchan software bug, not Dollchan.

**Comment by StephenLynx** [comment]
Is not a bug, old files don't have extensions.

**Comment by SthephanShinkufag** [comment]
What does mean "old files"? In old-api-lynxchan boards everything the same, links have extensions in href.

```html
<a class="imgLink" target="_blank" href="https://endchan.xyz/.media/229c397141dfbf7eb7422d6e1f840ec8-imagejpeg.jpg" data-filewidth="558" data-fileheight="600" data-filemime="image/jpeg">
    <img src="/.media/t_229c397141dfbf7eb7422d6e1f840ec8-imagejpeg" title="[Click] открыть по центру, [Ctrl+Click] в посте" width="237" height="255">
</a>
```

**Comment by StephenLynx** [comment]
Old files are pre 1.7 files. Most sites are not that old, but sites that migrated their content, migrated to 1.5 and then upgraded the data.

**Comment by ganego** [comment]
And when was that? Because I see files in new threads from time to time that are bugged.  
Is this because the file was not re-uploaded in this recent post but since it had the same MD5 as the same file uploaded sometime earlier, the old file was used?  
So basically we need a shell script that reads the magicbytes of all media files on the server and sets the extension accordingly to fix this and change the image file names in the database. Until then, it's bugged.

**Comment by StephenLynx** [comment]
>Is this because the file was not re-uploaded in this recent post but since it had the same MD5 as the same file uploaded sometime earlier, the old file was used?
Correct.

>Until then, it's bugged.
Everything is working as intended, is not bugged.

>Why would you even think in the first place, that saving files without extension makes any sense?
Extensions are not a standard, mimes are. I only added them back on 1.7 because too many people save files directly without using the link to download with the original filename, and desktops expect files to have extensions.

> And how do you get this?
I use the mime the client tells me. Which is as valid as the extension the client tells me. Then yes, I store this information, which is negligible when compared to the actual space used by the file itself.

**Comment by kr4ssi** [comment]
https://github.com/SthephanShinkufag/Dollchan-Extension-Tools/pull/1344

**Comment by SthephanShinkufag** [comment]
@kr4ssi

Show me a ling-live post with that kind of video. I must test it.

**Comment by kr4ssi** [comment]
I haven't seen it happening to videos too yet, for images it works

**Comment by ganego** [comment]
There is a bug with post dates.  
1) Open any thread in tab/browser A. 
2) Write some post. 
3) Open the same thread in a different tab/browser B.  
4) Times should be correct. In tab/browser B write a post.  
5) Refresh A -> Time is incorrect.


And while at it: Deleting posts does not work.

**Comment by ganego** [comment]
Admin set css for `video` `max-height: 90vh` in https://kohlchan.net/.static/css/posting.css
This will also affect dollchan making some videos not have full height (especially when min-webm-width is high).

Dollchan has to set `video { background: black; }` -> `video { background: black; max-height: 100%; }`

**Comment by kr4ssi** [comment]
*.webp doesn't show up in preview.

**Comment by ritz127** [comment]
kohlchan has a code in https://kohlchan.net/.static/js/posting.js which adds unix timestamp download links for each file, currently it's not working with dollchan. initial load adds all links correctly but this is ignore by dollchan's auto-updater

**Comment by SthephanShinkufag** [comment]
@ritz127 
I don't understand what do you mean. Show me a screenshot.

**Comment by SthephanShinkufag** [comment]
@StephenLynx 

Please help. I recently implemented generating md5 for images and posting with `fileMd5`, `fileMime` and `fileName` fields, with checking through `checkFileIdentifier.js` for existed md5 entries. If md5 is existed then I send `fileMd5`, `fileMime` and `fileName` but don't send a file content. Otherwise I send `fileMd5`, `fileMime` and `fileName` and a content for the file.
But it doesn't work in second case! If a md5 already existed then everything works, but if it does not existed - that image fails.
For example I want to post 3 images, first and third are already existed on server and the second is new, After posting I will have only first and third in my post. The second image will be ignored.

Dollchan request parameters fields are seems NO DIFFERENT from parameters without a Dollchan. But with Dollchan I have a problem! I’ve been trying to find a problem for several days, but still without result.

Here the link to test script..
https://dropmefiles.com/SvGqk

How to check:
1. Try to post a new image. Posting doesn't work.
2. Post it without Dollchan. Posting works.
3. Post it again with Dollchan. Posting works!

(!) If you post on lynxhub.com/test then post some text without Dollchan first, to pass captcha.
(!) Don't use Greasemonkey if you want to analyse for requuest parameters because network monitor in Firefox doesn't see Greasemonkey requests.
(!) This is a test script, don't use it for regular posting on boards.

**Comment by SthephanShinkufag** [comment]
@StephenLynx 

Please help. I recently implemented generating md5 for images and posting with `fileMd5`, `fileMime` and `fileName` fields, with checking through `checkFileIdentifier.js` for existed md5 entries. Also it is possible to post spoiler pictures.
So, if md5 is existed then I send `fileMd5`, `fileMime` and `fileName` but don't send a file content. Otherwise I send `fileMd5`, `fileMime` and `fileName` and a content for the file.
But it doesn't work in second case! If a md5 already existed then everything works, but if it does not existed on server - that image fails.
For example I want to post 3 images, first and third are already existed on server and the second is new, After posting I will have only first and third in my post. The second image will be ignored.

Dollchan request parameters fields are seems NO DIFFERENT from parameters without a Dollchan. But with Dollchan I have a problem! I’ve been trying to find a problem for several days, but still without result.

Here the link to test script..
https://github.com/SthephanShinkufag/Dollchan-Extension-Tools/raw/26a560815863d72cb260e6881bec342163a00123/src/Dollchan_Extension_Tools.es6.user.js

How to check:
1. Try to post a new image with Dollchan. Posting doesn't work.
2. Post it without Dollchan. Posting works.
3. Post it again with Dollchan. Posting works!

(!) If you post on lynxhub.com/test then post some text without Dollchan first, to pass captcha.
(!) Don't use Greasemonkey if you want to analyse for requuest parameters because network monitor in Firefox doesn't see Greasemonkey requests.
(!) This is a test script, don't use it for regular posting on boards.

**Comment by StephenLynx** [comment]
Where are you appending the file content?

Here?
if(isAppend) {
        data.append(name, val);
 }

If this is it, then you are missing the file name. Did you see how my code works? On api.js?

**Comment by SthephanShinkufag** [comment]
Nope. The problem is something else.

**Comment by SthephanShinkufag** [comment]
Also without 
```js
			data.append('fileMd5', md5Value);
			data.append('fileMime', mime);
			data.append('fileName', newFileName);
```

(as in stable script)
Posting of new images works.

**Comment by SthephanShinkufag** [comment]
1. Without additional fields - posting of new images works.
2. With additional fields and without content - posting of existed images by md5 works.
3. With additional fields and with content - posting of new images does not work.

**Comment by StephenLynx** [comment]
1: the engine is just using no-js upload
2: the engine is just referencing images based on the md5+mime
3: something is mismatched between the actual file and the metadata (md5, mime, name)

**Comment by StephenLynx** [comment]
Btw, are you sending the spoiler status for all of them?

**Comment by SthephanShinkufag** [comment]
Yes, `fileSpoiler` for every attached file. As I said, I compared requests parameters in the network monitor of my browser, and everything seemed to be the same.

**Comment by SthephanShinkufag** [comment]
`fileSpoiler` parameter is sending through the checkbox element in the postform, that created by Dollchan.

**Comment by StephenLynx** [comment]
No, is not that. I am talking about individual file spoiler information. Again. DID YOU READ MY API.JS CODE?

**Comment by SthephanShinkufag** [comment]
Compare the parameters of requersts. They are the same.

**Comment by StephenLynx** [comment]
Ah yeah, I misread it lmao.

**Comment by StephenLynx** [comment]
I dunno, I don't know how to use dollchan to test it out. But you should double check what you are sending. Either the md5 and mime of the file is not matching, or the length of one of the meta data is shorther. Try putting the new file before the existing files when you post and see if something changes.

**Comment by StephenLynx** [comment]
And btw, kc uses the extension to figure out mime instead of what the browser reports. So take that in consideration. If you look at their FE, you will see this method is changed there too.

**Comment by zedmike** [comment]
> There is a bug with post dates.
> 
> 1. Open any thread in tab/browser A.
> 2. Write some post.
> 3. Open the same thread in a different tab/browser B.
>    -> Times should be correct.
> 4. In tab/browser B write a post.
> 5. Refresh A -> Time is incorrect.
> 
> And while at it: Deleting posts does not work.

I always get a post date that is two hours behind the older post's dates.
Only new posts have the two hour behind issue, no matter how the "Time correction in posts*" setting and parser filter is changed.

**Comment by ritz127** [comment]
> @ritz127
> I don't understand what do you mean. I can't see any timestamp link. Show me a screenshot.

It is disabled by default, so you need to enable it in Settings > Other > UNIX timestamp filenames if you want to test it by yourself. Then enter a thread and wait until someone posts a picture, or just do it yourself in a different tab. 
It creates a link in this format: https://kohlkanal.net/.media/ce06487aa14e416ff0eae7096a070ca1-imagejpeg.jpg/1564801999367.jpg but the last part is not added to posts loaded after the original page load.

Most likely this is happening because posting.updateAllUnixFilenames() is not being called ( line 74 here https://gitgud.io/Tjark/KohlNumbra/blob/master/static/js/posting.js )
The same goes for posting.updateAllLocalTimes() which is probably causing the bug with post dates using UTC if loaded by dollchan's auto-update feature but local time otherwise.

[https://i.imgur.com/idgc6Ul.png](Here's a screenshot where you can see both bugs - the two posts were made one hour apart, not four)

**Comment by SthephanShinkufag** [comment]
Built-in native thread updater was not blocked by Dollchan because of CSP content security policy. Also youtube embedder did not show youtube thumbnails due to CSP.
`nocsp.kohlchan.net` does not have a blocking.

**Comment by ganego** [comment]
Add `ibDomains['kohlkanal.net'] = Kohlchan;`

**Comment by kr4ssi** [comment]
> Add `ibDomains['kohlkanal.net'] = Kohlchan;`

Shouldn't that usually work if I just put it under the line with kohlchan.net? Why doesn't it?

**Comment by ganego** [comment]
> 
> 
> > Add `ibDomains['kohlkanal.net'] = Kohlchan;`
> 
> Shouldn't that usually work if I just put it under the line with kohlchan.net? Why doesn't it?

It does.

**Comment by kr4ssi** [comment]
Under certain circumstances dollchan tries to create a new thread instead of replying to the existing one when quick-replying from the board index, I think only if before any thread got polled, but I'm not sure.

**Comment by kr4ssi** [comment]
There also is no quick-reply-button for posts invoked by an url-anchor.

**Comment by HuongChanoi** [comment]
Hello Team,

the addon doesn't seem to work with the paid Tempermonkey Extension of Safari with Kohlchan. It works fine with 4chan though.

It asks for permission but the button wont show up and none of the functions are visible.

**Comment by HuongChanoi** [comment]
Hello Team,

all posts that are submitted via the postmenu (reply-to function) are not working since the recent captcha update for kc a few days ago. Also the post date offset function doesn't seem to work either.

Kindly check up on the issue.

**Comment by kr4ssi** [comment]
Drawing doesn't work with dollchan activated

**Comment by ganego** [comment]
@SthephanShinkufag There are two really annoying bugs with KC left in DC. One of them stopping some people from using DC at all on KC.

1) When using a proxy (or Tor), KC will give you a captcha to solve every two days (or until you delete the captcha cookie). This captcha cannot be solved with DC enabled because the popup won't show.  
2) Deleting posts does not work, gives some error.

--------------------------
Info:
1) Can be tested by accessing KC via Tor. Either the normal URL or the onion URL (see kohlchan.net/links.html). Be aware that when using Tor, you will first have to wait for/solve the "Kohlcash" test. It's simply a script that brute forces some hash (nothing to do with bitcoin or so despite the name) in order to stop people from spamming. You only have to do this once (if you don't delete this cookie). This has nothing to do with the captcha problem and works fine. Just telling you so you won't be surprised.  
You can write test posts to some thread in /m/ (trash board).

2) Error:
 
```
Internal error:
TypeError: form.action.split is not a function
    at sendHTML5Post (Dollchan%20Extension%20Tools.user.js#1:15964:29)
    at html5Submit (Dollchan%20Extension%20Tools.user.js#1:9365:14)
    at addStuff/delBtn.onclick (Dollchan%20Extension%20Tools.user.js#1:14909:6)
```

Thank you

**Comment by ganego** [comment]
> Deleting posts does not work, gives some error.

Fix:
https://github.com/SthephanShinkufag/Dollchan-Extension-Tools/blob/082abae0ecd8ccbd678df7c52f804115110723fd/src/Dollchan_Extension_Tools.es6.user.js#L15961

`const task = form.action.split('/').pop();`
-> 
`const task = form.attributes.action.value.split('/').pop();`

Writing posts still works.

**Comment by Npaffen** [comment]
The "post menu" at KC is the button with the exclamation mark. It ordinarily shows a menu to report a post as "inappropriate" for example. I would really appreciate if this feature could be implemented. 

Blue Button on the right side :
http://prntscr.com/vfycpr
Post menu :
http://prntscr.com/vfydps

**Comment by pdftk** [comment]
If i try to post on Kohlchan with pictures or videos attached, i get an error message.

`Internal error:
TypeError: dc.body is undefined
    at getSubmitError (Dollchan%20Extension%20Tools.user.js?id=a0e200e5-601d-4d5c-b6f3-446ef4cd7035:9096:5)
    at checkUpload (Dollchan%20Extension%20Tools.user.js?id=a0e200e5-601d-4d5c-b6f3-446ef4cd7035:9114:13)`

Posting just text works fine.

**Comment by ganego** [comment]
> If i try to post on Kohlchan

.net or .onion?
Extension or script?
Version?

...

Install this: https://github.com/SthephanShinkufag/Dollchan-Extension-Tools/raw/master/src/Dollchan_Extension_Tools.es6.user.js

**Comment by pdftk** [comment]
I am using it on https://nocsp.kohlchan.net
I am using the Userscript (20.3.17.0) with Tampermonkey(4.11.6120) in Firefox(83.0).

**Comment by gruzefix** [comment]
>@ganego ganego mentioned this issue 5 days ago
>[Bug] Can't post pictures anymore on kohlchan.net #1389 

I did what was asked, namely
`
    Search for function checkUpload
    After const isDocument = data instanceof HTMLDocument; add: alert(isDocument);
    Before error = getSubmitError(_data); add a new line with: alert(1);
    Before error = getSubmitError(data); add a new line with: alert(2);
`

The popup returns "false".

I also used Violentmonkey instead of Tampermonkey as you suggested.

**Comment by ganego** [comment]
Ok the `false` should be correct and we now know where the error occurs exactly.  

Go into this function again and search for `data = JSON.parse(_data);` then add a new line above that with `console.log(_data);`.
Now open the devtools in FF (right-click somewhere on the page -> Inspect Element). Then click on "Console". Then write a post. There should be a new line. If everything was correct it should look like this: `{"status":"ok","data":12345}`. The number is the post number.  
Please tell what it says for a post without a picture that works and one where it does not work when you attach a picture.
When your post does not show up at all, at this point it should be the server rejecting your post and this reply will tell the reason (hopefully).

 Unfortunately as I said, I have no issues at all, so it's hard to investigate.

**Comment by gruzefix** [comment]
Text only:
`{"status":"ok","data":4587950}`

With an image it doesn't return anything, but the popup in the upper right corner gives out the following:
`TypeError: dc.body is undefined`
`    at getSubmitError (%20Dollchan%20Extension%20Tools.user.js#1:9105:5)`
`    at checkUpload (%20Dollchan%20Extension%20Tools.user.js#1:9124:13)`
`    at promise callback*_initAjaxPosting/this.form.onsubmit (%20Dollchan%20Extension%20Tools.user.js#1:8855:44)`

**Comment by ganego** [comment]
Strange.
Are you sure you put the line above the JSON? Does there at least show up an empty line? We are going to test that.

Remove the line you put there and now add the following below `const _data = (isDocument ? data.body.textContent : data).trim();`:


console.log(isDocument);
console.log(data);
console.log("---- data end");
console.log(_data);
console.log("---- _data end");

Post the result, with the maybe empty lines. Theoretically output should be `false` and `data` & `_data` should be identical.

Do you have JavaScript enabled for KC? Not that it should matter but who knows.

We cannot just remove the error handling, because it happens after the post was sent. So it's the server rejecting the post for whatever reason, otherwise it should have shown up at least at this point.

**Comment by ganego** [comment]
Updated version with several fixes for KC, support for KC onion URLs (including fixed image-search), support for m4v and ogg/opus files. You can now also click on the search button for an image/video and the name of it will be added as a quote.

https://github.com/ganego/Dollchan-Extension-Tools/blob/master/src/Dollchan_Extension_Tools.es6.user.js

**Comment by SthephanShinkufag** [comment]
Imageboards are shit. Why you are still visiting them?

**Comment by SthephanShinkufag** [comment]
>     1. When using a proxy (or Tor), KC will give you a captcha to solve every two days (or until you delete the captcha cookie). This captcha cannot be solved with DC enabled because the popup won't show.

Done. Dollchan asks for captcha now.

**Comment by ganego** [comment]
Another thing, when you preload images and then click on the file name, Firefox will download them with the blob's name/id instead of the correct file name. I fixed this in my build and now put it on github, see the last three commits (I made them in the correct files for you instead of the big script file): https://github.com/ganego/Dollchan-Extension-Tools/commits/master  

This will add an additional button in front of the file name, that will download the preloaded blob with the correct name.

**Comment by SthephanShinkufag** [comment]
@ganego 
Good idea, I will check tomorrow. Probably, it was necessary to create a new Issue, because this is not relate to the kohlchan. Or even a pull request.
Also, you can compile big script into modules, and vice versa:

1. Install [Git](https://git-scm.com/) and [Node.js](https://nodejs.org/) if not yet.
2. Install [Gulp](http://gulpjs.com/) using Git:
`npm rm -g gulp`
`npm install -g gulp-cli`
3. Using Git, go to Dollchan repo and update dependencies:
`cd Dollchan-Extension-Tools`
`npm install`
4. Then you can make both ES5 and ESNext userscript versions from [/src/modules](https://github.com/SthephanShinkufag/Dollchan-Extension-Tools/tree/master/src/modules) by running:
`gulp make`
5. Or you can make ESNext version only:
`gulp make:es6`
6. Also you can make separate source modules from your ESNext version:
`gulp make:modules`

**Comment by SthephanShinkufag** [comment]
@ganego 
https://github.com/SthephanShinkufag/Dollchan-Extension-Tools/commit/5809518748de1b58fb84c3e21dcaec191dd8c25e

![Image_links_02](https://user-images.githubusercontent.com/1059874/113620510-11d48080-9663-11eb-9ff0-41738fc31d40.jpg)
![Image_links_01](https://user-images.githubusercontent.com/1059874/113620523-14cf7100-9663-11eb-9c8b-356da24ffd80.jpg)

Now the searching button gives download links with names :
1. With original name
2. For webm - with the name obtained from the metadata (when enabled  `"Images"` > `"Load titles from WebM metadata"`)
2. With the number given by the board.

Metadata loaded from webm is now highlighted in yellow.

**Comment by ganego** [comment]
This is always enabled yes (did not test yet)?  

My use-case for the patch above is/was the following:
- Normal thread: Click on image name -> image gets downloaded with this name; all good.
- Thread with preload enabled: Problem: Click on image name downloads image with blob id; bad.
  - Solution: One button next to name only if picture was preloaded, so that **with one click** one can download the picture with the correct name like in a normal thread. So mainly I added an extra button for convenience so that when you download like 30 pictures, it's faster, but your solution is ok too for most people I guess. Although, now that I think about it, the extra button has the advantage that you can see when and for what pictures/videos preloading fails (which it does from time to time on KC).

"Name  from Metadata" is a nice idea.

**Comment by SthephanShinkufag** [comment]
No additional buttons!!!

**Comment by SthephanShinkufag** [comment]
> Thread with preload enabled: Problem: Click on image name downloads image with blob id; bad. 

Lol wut. This is krautchan-only bug maybe.

![Preload_names](https://user-images.githubusercontent.com/1059874/113628048-038b6200-966d-11eb-86da-a1b1fc531cd7.gif)

**Comment by SthephanShinkufag** [comment]
> Thread with preload enabled: Problem: Click on image name downloads image with blob id; bad. 

This is kohlchan bug. Due to unusual lynxchan qualities.

![Kohlchan_imagename_bug_01](https://user-images.githubusercontent.com/1059874/113629595-50703800-966f-11eb-82e5-e0b68e804d4a.jpg)
![Kohlchan_imagename_bug_02](https://user-images.githubusercontent.com/1059874/113629620-59f9a000-966f-11eb-83dc-3f97f38e23c6.jpg)

**Comment by ganego** [comment]
> This is not a blob.   

You did not preload the images...

**Comment by SthephanShinkufag** [comment]
@ganego

I show you the screenshot **before the preload**. Do you even know how the `download` attribute tag works? https://www.w3schools.com/tags/att_a_download.asp It allows you to download image with custom name. Download attribute on **your** screenshot has the value that Dollchan incorrectly takes from the href attrbiute before preload. This is not a blob! Compare the blob value and download value!

**Comment by SthephanShinkufag** [comment]
Oh god.

**Comment by SthephanShinkufag** [comment]
It looks like the kohlchan modifies the href via js after Dollchan preloads. 

```js
	preloadImages(data) {
...
		for(let i = 0; i < len; ++i) {
			const el = els[i];
			const imgLink = $parent(el, 'A');
			if(!imgLink) {
				continue;
			}
			let isRepToOrig = !!Cfg.openImgs;
			console.log(imgLink); // Test
```

As a result, the raw kohlchan link before preload:

```
<a class="imgLink" target="_blank" href="/.media/e8560b2a85ba98f802b5eed20d3e61c998c56c3ff07198a8bb957c8379aed85c.png" data-filewidth="988" data-fileheight="530" data-filemime="image/png"><img alt="" src="/.media/t_e8560b2a85ba98f802b5eed20d3e61c998c56c3ff07198a8bb957c8379aed85c" width="200" height="107"></a>
```

It doesn't contains original name, only lynxchan's inner name.
> /.media/e8560b2a85ba98f802b5eed20d3e61c998c56c3ff07198a8bb957c8379aed85c.png
But after kohlchan doing its scripts it have the correct href
> /.media/e8560b2a85ba98f802b5eed20d3e61c998c56c3ff07198a8bb957c8379aed85c.png/dl/Zeitplan.png

**Comment by SthephanShinkufag** [comment]
Done. As I said, Dollchan took the wrong source.
`.imgLink` (image container link) have just `/.media/e8560b2a85ba98f802b5eed20d3e61c998c56c3ff07198a8bb957c8379aed85c.png`.
Now the href will be taken from  `.originalNameLink` have `/.media/e8560b2a85ba98f802b5eed20d3e61c998c56c3ff07198a8bb957c8379aed85c.png/dl/Zeitplan.png`

**Comment by ganego** [comment]
Noice. Seems to work.

**Comment by kr4ssi** [comment]
Kohlchan does a hash-check before uploading a file, so the upload will get skipped if the file is up already.
This doesn't work with dollchan enabled, the file will get uploaded every time.

To reproduce:
-Upload a big file (dollchan may be enabled)
-Disable Dollchan
-Upload the same file again
-Upload gets skipped (And still posted)
-Enable Dollchan
-File gets uploaded again (And posted)

**Comment by kr4ssi** [comment]
When pointing on a link to an answered post on the board-index, the "Nr."-Link in this preview points to the painpage, like https://www.kohlchan.net/b/#4840205 (except for when this post is still visible on the index-page)
![grafik](https://user-images.githubusercontent.com/44404263/114306924-c6482980-9add-11eb-8858-a7cac0ae141a.png)

**Comment by kr4ssi** [comment]
![grafik](https://user-images.githubusercontent.com/44404263/114307032-23dc7600-9ade-11eb-84fe-258e65e9c7f5.png)

**Comment by kr4ssi** [comment]
Sometimes, when quick-replying from the board-index, a new thread is created instead of replying. I'm not sure how to reproduce yet, maybe related to the bug above.

**Comment by kr4ssi** [comment]
Tools for painting aren't visible in quick-reply

**Comment by SthephanShinkufag** [comment]
@ganego
> There is a bug with post dates. 
> 1) Open any thread in tab/browser A. 
> 2) Write some post. 
> 3) Open the same thread in a different tab/browser B.  -> Times should be correct. 
> 4) In tab/browser B write a post.  
> 5) Refresh A -> Time is incorrect.

@zedmike
> I always get a post date that is two hours behind the older post's dates.
Only new posts have the two hour behind issue, no matter how the "Time 
correction in posts*" setting and parser filter is changed.</task-lists></div>

@ritz127 
> kohlchan has a code in https://kohlchan.net/.static/js/posting.js which adds unix timestamp download links for each file, currently it's not working with dollchan. initial load adds all links correctly but this is ignore by dollchan's auto-updater.

> It is disabled by default, so you need to enable it in Settings > Other > UNIX timestamp filenames if you want to test it by yourself. Then enter a thread and wait until someone posts a picture, or just do it yourself in a different tab. 
It creates a link in this format: https://kohlkanal.net/.media/ce06487aa14e416ff0eae7096a070ca1-imagejpeg.jpg/1564801999367.jpg but the last part is not added to posts loaded after the original page load.
> Most likely this is happening because posting.updateAllUnixFilenames() is not being called ( line 74 here https://gitgud.io/Tjark/KohlNumbra/blob/master/static/js/posting.js )
The same goes for posting.updateAllLocalTimes() which is probably causing the bug with post dates using UTC if loaded by dollchan's auto-update feature but local time otherwise.
> [Here's a screenshot where you can see both bugs - the two posts were made one hour apart, not four](https://i.imgur.com/idgc6Ul.png)

Fixed. Many thanks to @konacha !

**Comment by konacha** [comment]
@SthephanShinkufag Something in [this update](2929813034e4819dbbf5ed5b85e6105cdf5fba91) broke the reverse search menu. It no longer shows on hovering. Not sure if it's related to the unix timestamp fix, because it still worked as of my last commit.

**Comment by SthephanShinkufag** [comment]
@konacha 
That was trying to add an ability to download with unixtime names through menu.

**Comment by konacha** [comment]
Due to recent refactoring some stuff stopped working (again ;_;) 
I had to block thumbs.js in my browser, otherwise Dollchan can't expand videos properly.
Posting gets stuck on "checking" forever. Posts are still submitted, but files are ignored.

**Comment by SthephanShinkufag** [comment]
@konacha 
Issues are observed even with 2019 script versions. I did not check it further. Also, in tampermonkey the problem is observed almost every time, and in greasemonkey it is very rare.

**Comment by SthephanShinkufag** [comment]
Probably the kohlchan admins changed something, and now it conflicts with Dollchan.

**Comment by konacha** [comment]
> Probably the kohlchan admins changed something, and now it conflicts with Dollchan.

That's what I meant, the refactor was on kohlchan's side
https://gitgud.io/Tjark/KohlNumbra/-/commits/master

It happened once before, when they added websockets and it broke Dollchan's autoupdate.
I haven't found any bugs other than posting (randomly works and stops working) and video thumbs

**Comment by SthephanShinkufag** [comment]
@konacha
Oh, I thought "Due to recent refactoring" was addressed to me.
Also will be absent from the mountain trip for the next couple of weeks, I will be glad if you find anything during this time. Thank you!

**Comment by zedmike** [comment]
Can't upload stuff on /int/

**Comment by zedmike** [comment]
Doesn't work on chrome when sending files.

**Comment by kr4ssi** [comment]
Click on image-expand links to /dl/, should be without /dl/.

**Comment by kr4ssi** [comment]
Impossible to post drawed image if dollchan enabled.

**Comment by pdftk** [comment]
It may help to know, that is possible to post with pictures attached, if the picture is very small. I posted yesterday on Kohlchan with dollchan enabled and a picture attached which had the size of 4,3 kb. The dimensions were 138x66 pixels.

**Comment by kr4ssi** [comment]
I was talking about drawed images, not uploaded images.

**Comment by kr4ssi** [comment]
```
async sendHTML5Post(form, data, needProgress, hasFiles) {
  if (unsafeWindow.oekaki?.expanded) {
    hasFiles = true;
    const dataURI = $('#wPaint')?.wPaint?.('image');
    const blob = unsafeWindow.oekaki?.dataURLtoBlob(dataURI);
    data.append('files', new File([blob], 'oekaki.png', { type: 'image/png' }));
  }
  return super.sendHTML5Post(form, data, needProgress, hasFiles);
}
```
This should work, but it doesn't. What am I doing wrong?

**Comment by kr4ssi** [comment]
https://github.com/SthephanShinkufag/Dollchan-Extension-Tools/pull/1410

**Comment by kr4ssi** [comment]
Why does Dollchan delete clickhandlers? How can I prevent it?

**Comment by kr4ssi** [comment]
Image-Preview seems broken in Lynxchan 2.7

**Comment by zedmike** [comment]
Hi where are the upgrades?

**Comment by SthephanShinkufag** [comment]
> Video-Preview is broken
> an issue with posting
> oekaki

Done, should work.

**Comment by kr4ssi** [comment]
The "Edit Image"-Button still doesn't work with dollchan enabled.

**Comment by kr4ssi** [comment]
Please make spoiler images possible with dollchan enabled.
![grafik](https://user-images.githubusercontent.com/44404263/177891569-3c963743-c990-49b8-be31-7b43d8dcc014.png)

**Comment by kr4ssi** [comment]
Canvas should close/clear after sending

Also deleting doesn't work with dollchan enabled

**Comment by kr4ssi** [comment]
Image expanding doesn't work anymore since last update.

**Comment by pdftk** [comment]
Since today, it does not work at all in Firefox. Neither the user script nor the addon.

**Comment by SthephanShinkufag** [comment]
@pdftk maybe you need to pass the cloudflare.com challenge at https://kohlchan.net/unlocked.html

**Comment by pdftk** [comment]
The cookie was already set from the challenge.  I thought it might be a incompatibility with another addon. Image expanding did not work, videos are ok. So i started a private window with only tampermonkey aloud. The image expanding worked, but videos wouldn't expand. They just played on the site how they natively would.
Anyway, thanks for your work.

**Comment by SthephanShinkufag** [comment]
@pdftk Ah, I thought your script didn't work at all. Fixed the video expansion.

Since this issue is too big and outdated, I'm closing it. Please create new issues for each question next time, so it will be easier for me to track them. It's very inconvenient to communicate and track what and where here.
